### PR TITLE
Allow adding plugin strokes to selection

### LINKED ISSUE #6028: Add selection operations to plugin API
### Is Your Feature Request Related to a Problem?

Currently there is no way manipulating selections from the plugin API. It would be convenient to:
- Create a selection from elements of the current layer
- Extend an existing selection with elements of the current layer
- Remove elements from the current selection and either delete them or release them to the current layer

### Feature Description

As a solution (and first idea) I suggest:
- element getters like `app.getStrokes` should include a pointer (lightuserdata in Lua) for every element, so they can be used for creating and manipulating selections
- element adders like `app.addStrokes` should return a table of pointers for every added element, so they can be used for creating and manipulating selections
- there should be API like `app.addToSelection(tbl)` that takes a table of pointers to elements on the current layer and make them part of the current selection (or create a new one, if there isn't already one)
~- there should be API like `app.removeFromSelection(tbl, mode)` that takes a table of pointers of elements in the current selection and either releases them to the current layer or removes them depending on the value of `mode`~
- there should be API like `app.clearSelection()` that releases all elements from the selection to the current layer and returns a table of pointers to the elements of the selection

### Alternative Solutions

_No response_

### Additional Context

This would solve #6002 by using `app.addToSelection` with the return value of `app.addStrokes`.

### PR DESCRIPTION
**Pull request by rolandlo**
Closes #6028, #6002.

- [x] modify `app.getStrokes` by adding references to return value
- [x] modify `app.addStrokes` by adding return value
- [x] modify `app.addSplines` by adding return value
- [x] modify `app.getTexts` by adding references to return value
- [x] modify `app.addTexts` by adding return value
- [x] modify `app.getImages` by adding references to return value
- [x] modify `app.addImages` by adding replacing return values by a table and nil with a ref
- [x] add `app.addToSelection`
- [x] add `app.clearSelection`

**Edit:** Removed add `app.removeFromSelection`, since it can be achieved by a combination of the other API functions.

### CODE REVIEW

**Comment by rolandlo** [comment]
Here's a demo for `app.addToSelection` with the following plugin code:
```lua
function initUi()
  app.registerUi({["menu"] = "Add some strokes to the selection", ["callback"] = "cb_stroke"});
end

function cb_stroke()
    local triangle = { 
      x = { 300.0, 500.0, 400.0, 300.0 }, 
      y = { 300.0, 300.0, 450.0, 300.0 }, 
      width = 1.5, 
      tool = "pen", 
      color = 0xff0000,
    }
    local square = { 
      x = { 300.0, 500.0, 500.0, 300.0, 300.0 }, 
      y = { 500.0, 500.0, 700.0, 700.0, 500.0 }, 
      width = 1.5, 
      tool = "pen", 
      color = 0x00ff00,
    }
    local refs = app.addStrokes { strokes = { triangle, square }, allowUndoRedoAction = "grouped" }
    app.refreshPage()
    app.addToSelection(refs)
end
```

https://github.com/user-attachments/assets/6e5666e3-c806-490f-9ece-ce3bf35d5920

**Comment by rolandlo** [comment]
Unfortunately in `app.addImages` we already have a return value, one per image, like:
```lua
nil, nil, "Error: Inserting the image abc.jpg failed", nil
```

with `nil` standing for success. It would be better to have a return value as a table like.

```lua
{
  userdata: 0x5f644c02c538, 
  userdata: 0x5f644c02c566, 
  "Error: Inserting the image abc.jpg failed", 
  userdata: 0x5f644c02c892
}
```
Not sure what to do with backwards compatibility.

**Comment by atticus-sullivan** [comment]
Table or multiple return arguments doesn't really make a difference I guess, from the lua side you can always call `table.pack`/`table.unpack` (except that `unpack`ing might not work correctly if there are `nil`s in the array). Anyhow the table must be used with `pairs` (instead of `ipairs`) because of the potential `nil` values.

About backwards compatibility, well one might argue `addImages` is not in a stable (?) release, it's just in the `nightly` release yet (as far as I can see).

**Comment by rolandlo** [comment]
@atticus-sullivan Can you take another look?

**Comment by atticus-sullivan** [comment]
Looks good to me. (Note: I didn't go and check all of the modifications regarding the lua stack index in `addImages` )

**Comment by rolandlo** [comment]
> ```lua
> local strokes = app.getStrokes("selection")
> ...
> app.clearSelection()
> app.addToSelection(refs)
> app.uiAction { action = "ACTION_DELETE" }
> ```
> 
> why deselect and then select the same elements again? Is this just since you assume the selection might have changed in the meantime?

It's only part of the elements (red strokes in the example).

**Comment by rolandlo** [comment]
> Looks good to me. (Note: I didn't go and check all of the modifications regarding the lua stack index in `addImages` )

I think I checked those carefully, since I dumped the stack at every step using https://stackoverflow.com/questions/59091462/from-c-how-can-i-print-the-contents-of-the-lua-stack

File: src/core/plugin/luapi_application.h [hunk 1]
```diff
@@ -688,6 +694,21 @@ static int handleUndoRedoActionHelper(lua_State* L, Control* control, const char
 }
 
 
+/**
+ * Helper function for API adding elements. Empties the stack and then pushes the pointers
+ * to all elements in the given vector into a table on the stack.
+ */
+static void refsHelper(lua_State* L, std::vector<Element*> elements) {
+    lua_settop(L, 0);
```
[the remark this answers is not in the document]
Comment (rolandlo) [reply src/core/plugin/luapi_application.h:702]: Fair point.

**Comment by atticus-sullivan** [comment]
> plugin code

Oh right, I completely missed that one

**Comment by atticus-sullivan** [comment]
> > Looks good to me. (Note: I didn't go and check all of the modifications regarding the lua stack index in `addImages` )
> 
> I think I checked those carefully, since I dumped the stack at every step using https://stackoverflow.com/questions/59091462/from-c-how-can-i-print-the-contents-of-the-lua-stack

That's an interesting reference. Maybe something we should document somewhere (but no idea where, as comment in the `luapi_application.h` file? :man_shrugging:)

**Comment by rolandlo** [comment]
I will take a look what existing plugins can be improved with the new API.

**Comment by rolandlo** [comment]
> I will take a look what existing plugins can be improved with the new API.

Done. Mostly it was only the ImageActions plugin. So from my side this PR is ready.

Merging in 72 hours if no objections are raised.

**Pull request merged**